packages/web-components/src/components/va-date-text-input/va-date-text-input.tsx

```diff
@@ -0,0 +1,279 @@
+      day ? numFormatter.format(dayNum) : ''
+    }`;
+
+    const daysForSelectedMonth = monthNum > 0 ? days[monthNum] : [];
+    const leapYear = checkLeapYear(yearNum);
+    // Check validity of date if invalid provide message and error state styling
+    if (this.customValidationBoolean) {
+      this.error = this.customValidationMessage;
+    } else if (
+      yearNum < minYear ||
+      yearNum > maxYear ||
+      monthNum < minMonths ||
+      monthNum > maxMonths ||
+      dayNum < minMonths ||
+      dayNum > daysForSelectedMonth.length ||
+      day === '' ||
+      month === '' ||
+      year === '' ||
+      (!leapYear && monthNum === 2 && dayNum > 28) ||
+      (this.required && !isFullDate(this.value))
+    ) {
+      this.error = 'Please provide a valid date';
+    } else {
+      this.error = '';
+    }
+    this.dateBlur.emit(event);
+  };
+
+  private handleDateChange = (event: Event) => {
+    const target = event.target as HTMLInputElement;
+    let [currentYear, currentMonth, currentDay] = (this.value || '').split('-');
+    if (target.classList.contains('input-month')) {
+      currentMonth = target.value;
+    }
+    if (target.classList.contains('input-day')) {
+      currentDay = target.value;
+    }
+    if (target.classList.contains('input-year')) {
+      currentYear = target.value;
+    }
+
+    this.value = `${currentYear}-${currentMonth ? currentMonth : ''}-${
+      currentDay ? currentDay : ''
+    }`;
+
+    // This event should always fire to allow for validation handling
+    this.dateChange.emit(event);
+  };
+
+  private handleDateKey = (event: KeyboardEvent) => {
+    // Allow 0-9, Arrow Left, Arrow Right, Delete, Backspace and Tab to clear data or move to next field
+    const validKeys = [
+      '0',
+      '1',
+      '2',
+      '3',
+      '4',
+      '5',
+      '6',
+      '7',
+      '8',
+      '9',
+      'Backspace',
+      'ArrowRight',
+      'ArrowLeft',
+      'Tab',
+      'Delete',
+    ];
+    if (validKeys.indexOf(event.key) < 0) {
+      event.preventDefault();
+      return false;
+    }
+  };
+
+  /**
+   * Whether or not an analytics event will be fired.
+   */
+  @Prop() enableAnalytics: boolean;
+
+  /**
+   * The event used to track usage of the component. This is emitted when an
+   * input value changes and enableAnalytics is true.
+   */
+  @Event({
+    eventName: 'component-library-analytics',
+    composed: true,
+    bubbles: true,
+  })
+  componentLibraryAnalytics: EventEmitter;
+
+  connectedCallback() {
+    i18next.on('languageChanged', () => {
+      forceUpdate(this.el);
+    });
+  }
+
+  disconnectedCallback() {
+    i18next.off('languageChanged');
+  }
+
+  render() {
+    const {
+      required,
+      label,
+      name,
+      error,
+      handleDateBlur,
+      handleDateChange,
+      handleDateKey,
+      value,
+    } = this;
+
+    const [year, month, day] = (value || '').split('-').map(val => val);
```

Comment:
This is a two-parter. It looks like the three consts here would be strings or `undefined`. If the map function was something like `val => val || ''`...

Reply:
See comment below